### PR TITLE
[WIP | Benchmarking] Optimize bbolt memory usage

### PR DESCRIPTION
**Pull request by turan18**
**Issue #, if available:**

Resolves: #969 

**Description of changes:**

**Testing performed:**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by austinvazquez** [comment]
@turan18 can you document if there is any performance trade-offs to chunked insertion?

File: metadata/reader.go [hunk 1]
```diff
@@ -671,3 +686,13 @@ func (r *reader) NumOfNodes() (i int, _ error) {
 	}
 	return
 }
+
+// partition partitions a slice of type T into chunks of size chunkSize
+func partition[T any](data []T, chunkSize int) [][]T {
```
Comment (austinvazquez) [inline metadata/reader.go:691]: Can we add some unit tests that show this is chunking correctly? Also we should either add a unit test or documentation that chunkSize never be zero or this function will segfault.

File: metadata/reader.go [hunk 2]
```diff
@@ -53,7 +53,13 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
-// reader stores filesystem metadata parsed from ztoc to metadata DB
+const (
+	// bboltInsertionChunkSize roughly determines the maximum number of insertions to
+	// bbolt per transaction.
+	bboltInsertionChunkSize = 5000
```
Comment (austinvazquez) [inline metadata/reader.go:59]: Can we document why 5000? Even a line stating "manually tuned and gave the best performance" could give context in a few years time.

File: metadata/db.go [hunk 3]
```diff
@@ -325,16 +332,18 @@ func writeMetadataEntry(md *bolt.Bucket, m *metadataEntry) error {
...
-		}
-		if err := md.Put(bucketKeyChildName, []byte(firstChildName)); err != nil {
-			return fmt.Errorf("failed to put name first child %q: %w", firstChildName, err)
-		}
 		if len(m.children) > 1 {
+			// Sort children by base name.
+			keys := make([]string, len(m.children))
+			for k := range m.children {
+				keys = append(keys, k)
+			}
```
Comment (austinvazquez) [inline metadata/db.go:340]: If we want to preallocate, then it should be `keys := make([]string, 0, len(m.children)`. Right now we are using twice the memory.

The current form will create the container with size len(m.children) and fill with zero values. i.e. empty strings. This results in the empty keys check below. That can be removed with this fix.

File: metadata/db.go [hunk 4]
```diff
@@ -325,16 +332,18 @@ func writeMetadataEntry(md *bolt.Bucket, m *metadataEntry) error {
...
+			sort.Slice(keys, func(i, j int) bool {
+				return bytes.Compare([]byte(m.children[keys[i]].base), []byte(m.children[keys[j]].base)) < 0
+			})
 			var cbkt *bolt.Bucket
-			for k, c := range m.children {
-				if k == firstChildName {
+			for _, key := range keys {
+				if key == firstChildName || key == "" {
```
Comment (austinvazquez) [inline metadata/db.go:346]: `key == ""` check can be removed with above fix.

**Review by austinvazquez — CHANGES_REQUESTED** [review]

File: metadata/db.go (hunk 3 above)
Comment (turan18) [reply metadata/db.go:340]: Ah, right. I forgot the capacity is doubled once we hit the length.

File: metadata/reader.go (hunk 1 above)
Comment (turan18) [reply metadata/reader.go]: Sure, I can add unit tests. This was taken from the Go Wiki: https://go.dev/wiki/SliceTricks#batching-with-minimal-allocation.

**Comment by turan18** [comment]
> @turan18 can you document if there is any performance trade-offs to chunked insertion?

Added the pull performance benchmarks to the description.

File: metadata/db.go [hunk 5]
```diff
@@ -162,36 +170,16 @@ func getMetadataBucketByID(md *bolt.Bucket, id uint32) (*bolt.Bucket, error) {
...
-		if v.val != 0 {
-			val, err := dbutil.EncodeInt(v.val)
-			if err != nil {
-				return err
-			}
-			if err := b.Put(v.key, val); err != nil {
-				return err
-			}
-		}
+// writeMetadataEntry writes node metadata to the appropriate node sub-bucket inside the top-level nodes bucket.
```
Comment (austinvazquez) [inline metadata/db.go:173]: nit: writeNodeEntry*

**Review by austinvazquez — DISMISSED** [review]
Minor nit, but changes look good to me. Thanks for all the detail in the PR description.

**Review by austinvazquez — APPROVED** [review]

**Review by sondavidb — APPROVED** [review]
